### DISCUSSION TITLE
How should hot-key presets work?

### DISCUSSION THREAD

**Discussion by macmarrum** [freeplane/freeplane#1267] [General]
In Freeplane, a set of hot keys can be saved and restored:
- `Tools->Hot key presets->Save hot keys…`
- `Tools->Hot key presets->Load`

I've been writing the documentation page [Hot Keys and Beyond](https://docs.freeplane.org/user-documentation/hot-keys-and-beyond.html) and ran into the problem of hot keys working in an unexpected way:
- #1250 
- #1252

Encouraged by Dimitry, this is to gather opinions from the community.

Here's how I see the topic, represented in a more structured way of a User Story and Acceptance Criteria - a method often used during working on software changes.

Please feel free to provide your opinion about the below or new requirements in any form you like.

<h1>User Story</h1><table>
<colgroup><col /><col /></colgroup>
<tbody>
<tr>
<th>AS</th>
<td>a Freeplane user</td>
</tr>
<tr>
<th>I WANT</th>
<td>Freeplane to save and restore my hot-key presets</td>
</tr>
<tr>
<th>SO THAT</th>
<td>I can use the same hot key for different actions, depending on the context (represented by a hot-key preset)</td>
</tr>
</tbody>
</table> 
<h1>Acceptance Criteria</h1><table>
<colgroup><col /><col /></colgroup>
<tbody>
<tr>
<th>AC1</th>
<td><b>Once a saved hot-key preset is restored, hot keys point to the actions/icons assigned to them when the preset was saved.</b><br />
<br />
User installs Freeplane and a new profile is created when Freeplane is started.<br />
<i>Note: the default (fallback) hot-key assignment is available, i.e. <code>Ctrl+X</code> activates <code>Edit->Cut</code>.</i><br />
User saves the hot-key preset as <code>default</code>.<br />
User assigns <code>Ctrl+X</code> to the icon <code>Icons->Signs->Not OK</code> and saves the preset as <code>icons</code>.<br />
 &rArr; <code>Ctrl+X</code> activates <code>Icons->Signs->Not OK</code>, which inserts the icon <code>Not OK</code>.<br />
User changes the hot-key preset back to <code>default</code>, i.e. <code>Tools->Hot key presets->Load->default</code>.<br />
 &rArr; <code>Ctrl+X</code> activates <code>Edit->Cut</code>. </td>
</tr>
<tr>
<th>AC2</th>
<td>Once a saved hot-key preset is restored, hot keys point to the actions/icons assigned to them when the preset was saved.<br /><b>In case of dynamic actions (e.g. user styles), their hot keys are dormant until the actions appear</b><br />
<br />
User installs Freeplane and a new profile is created when Freeplane is started.<br />
<i>Note: the default (fallback) hot-key assignment is used, i.e. <code>Ctrl+I</code> activates <code>Format->Node core->Italic</code>.</i><br />
User saves the hot-key preset as <code>default</code>.<br />
User creates a custom style <code>Incredibly useful</code> and saves the mind map as <code>useful.mm</code>.<br />
User assigns <code>Ctrl+I</code> to <code>Format->Style->Incredibly useful</code> and saves the hot-key preset as <code>useful</code>.<br />
User restores the hot-key profile <code>default</code>.<br />
 &rArr; <code>Ctrl+I</code> activates <code>Format->Node core->Italic</code>.<br />
User closes <code>useful.mm</code> and opens a new mind map, which lacks the style <code>Incredibly useful</code>.<br />
User restores the hot-key profile <code>useful</code>.<br />
 &rArr; <code>Ctrl+I</code> triggers no action.<br />
User opens <code>useful.mm</code>.<br />
 &rArr; <code>Ctrl+I</code> activates <code>Format->Style->Incredibly useful</code>, which assigns the style <code>Incredibly useful</code>. </td>
</tr>
<tr>
<th>AC3</th>
<td><b>Once a hot key is unassigned, it goes back to the fallback (default), if such exists.</b><br />
<br />
User installs Freeplane and a new profile is created when Freeplane is started.<br />
<i>Note: the default (fallback) hot-key assignment is available, i.e. <code>Ctrl+X</code> activates <code>Edit->Cut</code>.</i><br />
User assigns <code>Ctrl+X</code> to the icon <code>Icons->Signs->Not OK</code>.<br />
 &rArr; <code>Ctrl+X</code> activates <code>Icons->Signs->Not OK</code>, which inserts the icon <code>Not OK</code>.<br />
User unassigns (clears) <code>Ctrl+X</code> from the icon <code>Icons->Signs->Not OK</code>.<br />
 &rArr; <code>Ctrl+X</code> activates <code>Edit->Cut</code>. </td>
</tr>
<tr>
<th>AC4</th>
<td><b>Hot keys restored from a preset in a running Freeplane remain assigned to the same action/icon after a restart</b><br />
<br />
User installs Freeplane and a new profile is created when Freeplane is started.<br />
User assigns <code>Ctrl+T</code> to <code>Tools->Show past reminders…</code><br />
User saves the hot-key preset as <code>past reminders</code>.<br />
User assigns <code>Ctrl+T</code> to <code>Tools->Manage time…</code><br />
User saves the hot-key preset as <code>manage time</code>.<br />
User restores the hot-key preset <code>past reminders</code>.<br />
 &rArr; <code>Ctrl+T</code> activates <code>Tools->Show past reminders…</code>.<br />
User quits Freeplane and starts it again.<br />
 &rArr; <code>Ctrl+T</code> activates <code>Tools->Show past reminders…</code>. </td>
</tr>
</tbody>
</table>

**Comment by euu2021** [comment]
I will comment about the item AC3, because I was assigning some hotkeys today, and got surprised with the behavior.

In my mind, FP had this fallback behavior, hadn't it?

Anyway, for the maximum clarity for the user, I suggest that the assignment dialog has also the option "Default". Like this:
![image](https://github.com/freeplane/freeplane/assets/77707706/5530d727-553e-4e7a-89ea-51222e19ab66)

This makes clear that clear means clear. Otherwise, the user doesn't know if, leaving it empty will revert to default or not.

**Reply by euu2021** [reply]
now, to make it even more clear, a message showing that no hotkey is assigned:

![image](https://github.com/freeplane/freeplane/assets/77707706/1daab20d-2bcc-4d2e-b8ef-75602f70efa6)

**Reply by euu2021** [reply]
this is even better: the user is presented with the Default hotkey, the currently assigned, and the new one; with the options to Clear and to revert to Default:
![image](https://github.com/freeplane/freeplane/assets/77707706/18216ba4-9f0c-4198-a4e5-e8ebf1e8e0a2)

**Reply by euu2021** [reply]
One more thing that I noticed: this same dialog could warn the user if this key is already assigned to another action:

![image](https://github.com/freeplane/freeplane/assets/77707706/642fb4c1-6523-43b5-b7ff-88064aa42e5e)

Currently, this warning only comes on the next dialog.

**Comment by EdoFro** [comment]
Hi @macmarrum,
I appreciate your effort, however I don't usually change my hotkey settings or save different versions. So I guess I am not a very relevant opinion in this case.
The only case where I can imagine making use of this functionality would be if two people were working on the same computer on the same map and swapping who is in front of the keyboard.
In that case each one could have their own keyboard configuration and could go back and forth between them so that the work would be comfortable and fluid. 

I understand the cases you raise, but it is not clear to me what happens with a hotkey that is assigned to a command in one of the profiles and not in the other.
For example, if in one of the profiles "ctrl+shift+g" creates a new floating node and in the other profile that key combination is not used and that command does not have an associated hotkey.
In that case, when changing the profile, is that hotkey kept because it does not conflict with the new profile or is it deleted because it is not included in the profile?

**Reply by macmarrum** [reply]
> I understand the cases you raise, but it is not clear to me what happens with a hotkey that is assigned to a command in one of the profiles and not in the other.
> For example, if in one of the profiles "ctrl+shift+g" creates a new floating node and in the other profile that key combination is not used and that command does not have an associated hotkey.
> In that case, when changing the profile, is that hotkey kept because it does not conflict with the new profile or is it deleted because it is not included in the profile?

This is a good point, @EdoFro. 
The way I see it, a preset should represent a state, including the absence of an assignment.
In your example, `Ctrl+Shift+G` should be removed, because it wasn't assigned when the preset was saved.
I'll update my Acceptance Criteria.

**Reply by EdoFro** [reply]
and the shortcuts added by an add-on?
should they be deleted if loading a state saved before the add-on installation?

**Reply by EdoFro** [reply]
I think maybe would be better that different saved states replace each other when loading but not deleting hotkey that have no conflict. 
AND:
a “Reset to default” command that deletes all and only sets the standard hotkeys from Freeplane and addons. 

this way:
- if you want to change only some of the hotkeys, then you load another state
- if you want to change all, than you can reset to default and then load the desired state. 

just a thought 
BR
edo

**Reply by EdoFro** [reply]
>… I don't usually change my hotkey settings or save different versions. So I guess I am not a very relevant opinion in this case.

please remember this. My opinion may not apply at all. 😬

**Reply by macmarrum** [reply]
> I think maybe would be better that different saved states replace each other when loading but not deleting hotkey that have no conflict.
AND:
a “Reset to default” command that deletes all and only sets the standard hotkeys from Freeplane and addons.
>
>this way:
>
> - if you want to change only some of the hotkeys, then you load another state
> - if you want to change all, than you can reset to default and then load the desired state.

In other words, two options when restoring a preset:
1. update existing and/or add new from preset
2. clear existing then add all from preset

I like this idea :bulb::+1:

**Comment by sifran-github** [comment]
In my opinion:
When saving keys, it should save all the keys that are currently configured and are not the default.
When loading keys, it should load: the default keys + override with the loaded keys in the file.

The implementation of not removing non-default keys when loading other key configurations is not what I expect when I load key configurations. It makes a mix of several configurations, which is not what I expect when I load something.

If it is very important for users to mix several configuration, I would expect an additional option of "Merge with current" but to me it is not needed, it will just confuse users, and... actually can be done manually by a simple text editor.

In this post I'm talking about general menu keys. I don't know what "hotkeys assigned to styles" is or how to use it.

**Reply by macmarrum** [reply]
@sifran-github Thanks for you feedback.
Could you please have a look if [this idea](https://github.com/freeplane/freeplane/discussions/1267#discussioncomment-6405195) covers what you described?

> In other words, two options when restoring a preset:
> 1. update existing and/or add new from preset
> 2. clear existing then add all from preset

**Reply by sifran-github** [reply]
@macmarrum if I understand correctly, we described the same functionality.

Conceptually, for me the "clear existing then add all from preset" is just called "load". In any app, when you "load" a file, it doesn't merge it with the existing file... it just loads it. for example, If you watch a document in Microsoft Word and you load another file... it opens the loaded file as it is, not merging it. I'm used to it also in relation to loading some configuration, like in this situation.

This is why I chose the words (1) Load and (2) Merge to describe these functions.
